src/mincraft/mincraft.cc

```diff
@@ -0,0 +1,1210 @@
+{
+    auto formats = std::to_array<vkwrap::SwapchainReqsBuilder::WeightFormat>(
+        { { .property = { vk::Format::eB8G8R8A8Unorm, vk::ColorSpaceKHR::eSrgbNonlinear }, .weight = 0 },
+          { .property = { vk::Format::eR8G8B8A8Unorm, vk::ColorSpaceKHR::eSrgbNonlinear }, .weight = 0 } } );
+
+    const auto weight_immediate = ( uncapped_always ? 0 : 100 );
+    auto modes = std::to_array<vkwrap::SwapchainReqsBuilder::WeightMode>(
+        { { .property = vk::PresentModeKHR::eFifo, .weight = 50 }, //
+          { .property = vk::PresentModeKHR::eMailbox, .weight = weight_immediate },
+          { .property = vk::PresentModeKHR::eImmediate, .weight = weight_immediate } } );
+
+    auto requirement_builder = vkwrap::SwapchainReqsBuilder{};
+    requirement_builder.withMinImageCount( k_max_frames_in_flight )
+        .withFormats( formats )
+        .withModes( modes )
+        .withSurface( surface );
+
+    return requirement_builder.make();
+}
+
+vkwrap::Swapchain
+createSwapchain(
+    vk::PhysicalDevice physical_device,
+    vk::Device logical_device,
+    vk::SurfaceKHR surface,
+    vkwrap::Queue graphics_queue,
+    vkwrap::Queue present_queue,
+    const vkwrap::SwapchainReqs& swapchain_requirements )
+{
+    auto min_image_count = std::max(
+        swapchain_requirements.getMinImageCount(),
+        physical_device.getSurfaceCapabilitiesKHR( surface ).minImageCount );
+
+    auto swapchain_builder = vkwrap::SwapchainBuilder{};
+    swapchain_builder.withQueues( std::array{ graphics_queue, present_queue } )
+        .withImageExtent( vkwrap::getSurfaceExtent( physical_device, surface ) )
+        .withSurface( surface )
+        .withMinImageCount( min_image_count )
+        .withImageUsage( vk::ImageUsageFlagBits::eColorAttachment )
+        .withPreTransform( vk::SurfaceTransformFlagBitsKHR::eIdentity )
+        .withCompositeAlpha( vk::CompositeAlphaFlagBitsKHR::eOpaque )
+        .withClipped( false )
+        .withOldSwapchain( {} );
+
+    auto new_swapchain = swapchain_builder.make( logical_device, physical_device, swapchain_requirements );
+
+    return new_swapchain;
+}
+
+struct FrameSyncPrimitives
+{
+    vk::UniqueSemaphore image_availible_semaphore;
+    vk::UniqueSemaphore render_finished_semaphore;
+    vk::UniqueFence in_flight_fence;
+};
+
+struct FrameRenderingInfos
+{
+    std::vector<FrameSyncPrimitives> sync_primitives;
+    std::vector<vk::UniqueCommandBuffer> imgui_command_buffers;
+    std::vector<vkwrap::Buffer> uniform_buffers;
+};
+
+FrameRenderingInfos
+createRenderInfos(
+    vk::Device logical_device,
+    vkwrap::CommandPool& command_pool,
+    ranges::range auto&& queues,
+    vkwrap::Mman& manager )
+{
+    FrameRenderingInfos frame_render_info;
+
+    auto frames = k_max_frames_in_flight;
+    auto buffer_builder = vkwrap::BufferBuilder{};
+    buffer_builder.withQueues( queues );
+
+    for ( [[maybe_unused]] auto i = uint32_t{ 0 }; i < frames; ++i )
```

Comment:
I think [[maybe_unused]] is useless in this context, only disturb reading. Also i prefer to write `uint32_t i = 0;` rather then `auto i = uint32_t { 0 };`